src/webserver/WsBase.h

```diff
@@ -383,29 +366,57 @@ class WsSession : public WsBase,
 
         BOOST_LOG_TRIVIAL(debug)
             << __func__ << " received request from ip = " << *ip;
-        auto sendError = [&](auto&& msg) {
-            boost::json::object response;
-            response["error"] = std::move(msg);
-            std::string responseStr = boost::json::serialize(response);
 
-            BOOST_LOG_TRIVIAL(trace) << __func__ << " : " << responseStr;
+        auto sendError = [this, ip](
+                             auto error,
+                             const boost::json::value& id,
+                             boost::json::object& request) {
+            auto e = RPC::make_error(error);
+
+            if (!id.is_null())
+            {
+                e["id"] = id;
+            }
+            e["request"] = request;
 
+            auto responseStr = boost::json::serialize(e);
+            BOOST_LOG_TRIVIAL(trace) << __func__ << " : " << responseStr;
             dosGuard_.add(*ip, responseStr.size());
             send(std::move(responseStr));
         };
+
+        boost::json::value raw = [](std::string const&& msg) {
+            try
+            {
+                return boost::json::parse(msg);
+            }
+            catch (std::exception&)
+            {
+                return boost::json::value{nullptr};
+            }
+        }(std::move(msg));
+
+        if (!raw.is_object())
+            return sendError(
+                RPC::Error::rpcINVALID_PARAMS, nullptr, raw.as_object());
+        boost::json::object request = raw.as_object();
+
+        auto id = request.contains("id") ? request.at("id") : nullptr;
+
         if (!dosGuard_.isOk(*ip))
         {
-            sendError("Too many requests. Slow down");
+            sendError(RPC::Error::rpcSLOW_DOWN, id, request);
         }
         else
         {
             if (!queue_.postCoro(
-                    [m = std::move(msg), shared_this = shared_from_this()](
+                    [shared_this = shared_from_this(), &request, id](
```

Comment:
I think this is actually unsafe. We capture `request` by reference, but then post to the job queue, after which this function returns and `request` goes out of scope. What do we need to parse the request for here? To get the id? Which we only need if we fail the DosGuard check? I think we can move the DosGuard check inside `handle_request` and send the error from there, and not have to parse the request here.

Reply:
Sounds good, I'll fix it

Reply:
This should fix the unsafe behavior
```cpp
[shared_this = shared_from_this(), request = std::move(request), id];
```

Reply:
The unsafe behavior is we capture `request` as a reference (just a memory address of `request`), which then goes out of scope (postCoro may be congested and take a while to handle the request). 

During the time that the request waits in the queue, `request` goes out of scope, and that reference points to memory that is now no longer allocated.